### DISCUSSION TITLE
Passing an id to a code cell

### DISCUSSION THREAD

**Discussion by andrewpbray** [quarto-dev/quarto-cli#10229] [Q&A] [answered]
### Question

Is there a way to pass a pandoc id to the top level block that is created by a code cell (the `Div` with a class `cell`)?

### Examples

When doing a crossref, you can the `label` key in a code along with the `fig-` prefix. That passes the value for `label` as the id of the `Image` that is created by the cell.

````qmd
---
format: native
---

```{r}
#| label: fig-histplot

hist(rnorm(5))
```
````

```
Pandoc
  Meta { unMeta = fromList [] }
  [ Div
      ( "" , [ "cell" ] , [] )
      [ CodeBlock
          ( "" , [ "r" , "cell-code" ] , [] ) "hist(rnorm(5))"
      , Div
          ( "" , [ "cell-output-display" ] , [] )
          [ Plain
              [ Image
                  ( "fig-histplot" , [] , [] )
                  []
                  ( "Untitled_files/figure-native/fig-histplot-1.png"
                  , ""
                  )
              ]
          ]
      ]
  ]
```

However, if you try to use that logic to generically assign an id to the enclosing `Div` with class `cell`, no luck:

````qmd
---
format: native
---

```{r}
#| label: histplot

hist(rnorm(5))
```
````
```
Pandoc
  Meta { unMeta = fromList [] }
  [ Div
      ( "" , [ "cell" ] , [] )
      [ CodeBlock
          ( "" , [ "r" , "cell-code" ] , [] ) "hist(rnorm(5))"
      , Div
          ( "" , [ "cell-output-display" ] , [] )
          [ Para
              [ Image
                  ( "" , [] , [] )
                  []
                  ( "Untitled_files/figure-native/histplot-1.png" , "" )
              ]
          ]
      ]
  ]
```
Note that not only is a non-prefixed label not propagated into the AST, there is a filter that scrubs that attribute from the block. 

### The behavior that I'd like to see

Fundamentally, I'd like to be able to create an AST that looks like

```
Pandoc
  Meta { unMeta = fromList [] }
  [ Div
      ( "histplot" , [ "cell" ] , [] )
      [ CodeBlock
          ( "" , [ "r" , "cell-code" ] , [] ) "hist(rnorm(5))"
      , Div
          ( "" , [ "cell-output-display" ] , [] )
          [ Para
              [ Image
                  ( "" , [] , [] )
                  []
                  ( "Untitled_files/figure-native/histplot-1.png" , "" )
              ]
          ]
      ]
  ]
```
where the enclosing `Div` of class `cell` that is generated by a code cell can be passed an id.

One approach to implementation would be to assign the value of `label` in this manner if it doesn't use one of the protected prefixes. Another second would be to create a separate key for this purpose. A third would be to pass `label: histplot` as an attribute on that enclosing `Div` so that I can move it from an attribute to an id using a filter.

Thanks!

```bash
$ quarto --version
1.5.47
```

**Comment by jimjam-slam** [comment]
I also note that `id` seems to work the same way as `label`, but `classes` targets the `cell`:

````
---
format: native
---

```{r}
#| id: fig-histplot
#| classes: my-histplot

hist(rnorm(5))
```
````

```
Pandoc
  Meta { unMeta = fromList [] }
  [ Div
      ( "" , [ "cell" , "my-histplot" ] , [] )
      [ CodeBlock
          ( "" , [ "r" , "cell-code" ] , [] ) "hist(rnorm(5))"
      , Div
          ( "" , [ "cell-output-display" ] , [] )
          [ Plain
              [ Image
                  ( "fig-histplot" , [] , [] )
                  []
                  ( "id_files/figure-native/fig-histplot-1.png" , "" )
              ]
          ]
      ]
  ]
```

I'm assuming that the ID moving to the output is part of making crossrefs work in Quarto, but it does complicate our use case, @andrewpbray!

**Comment by cderv** [comment, marked as answer]
quite a complex topic today - we would need to think of the best way to allow such addition. 

The `Div` of class `cell` is really something from quarto, with quite some internal usage. 

The two mechanism in Quarto are the following: 

- `classes` allows to set classes to Div of class cell. 
- Any unknown yaml option will be set as attributes to this Div too.  (see https://github.com/quarto-dev/quarto-cli/issues/6253)

an id of the div is quite specific, and there is no mechanism for id. 

This means you could do 

````markdown
---
format: native
---

```{r}
#| custom-id: '#histplot'

hist(rnorm(5))
```
````

````
Pandoc
  Meta { unMeta = fromList [] }
  [ Div
      ( "" , [ "cell" ] , [ ( "custom-id" , "#histplot" ) ] )
      [ CodeBlock
          ( "" , [ "r" , "cell-code" ] , [] ) "hist(rnorm(5))"
      , Div
          ( "" , [ "cell-output-display" ] , [] )
          [ Para
              [ Image
                  ( "" , [] , [] )
                  []
                  ( "test_files/figure-native/unnamed-chunk-1-1.png"
                  , ""
                  )
              ]
          ]
      ]
  ]
````

Then a Lua filter to catch `custom-id` attributes and move it as id of the block would work. 

> I also note that id seems to work the same way as label

I can't find it, but I wonder if this is not an alias... 🤔 

### Specificity of knitr engine.

**knitr** has a concept of `class.*` and `attr.*` options ([knitr options](https://yihui.org/knitr/options/#text-output)) 

For Quarto, they will work to pass to `Codeblock` part. 

If you try for example 
````markdown
---
format: native
---

```{r}
#| attr-source: '#histplot'

hist(rnorm(5))
```
````

````
Pandoc
  Meta { unMeta = fromList [] }
  [ Div
      ( "" , [ "cell" ] , [] )
      [ CodeBlock
          ( "histplot" , [ "r" , "cell-code" ] , [] ) "hist(rnorm(5))"
      , Div
          ( "" , [ "cell-output-display" ] , [] )
          [ Para
              [ Image
                  ( "" , [] , [] )
                  []
                  ( "test_files/figure-native/unnamed-chunk-1-1.png"
                  , ""
                  )
              ]
          ]
      ]
  ]
````



`attr.output` does work in some cases with evaluated content for the fenced code block created for outputs
````markdown
---
format: native
---

```{r}
#| attr-output: '#histplot'

1 + 1
```
````
````
Pandoc
  Meta { unMeta = fromList [] }
  [ Div
      ( "" , [ "cell" ] , [] )
      [ CodeBlock ( "" , [ "r" , "cell-code" ] , [] ) "1 + 1"
      , Div
          ( "" , [ "cell-output" , "cell-output-stdout" ] , [] )
          [ CodeBlock ( "" , [ "#histplot" ] , [] ) "[1] 2" ]
      ]
  ]
````
Like in **knitr** directly. No application for plot output for example. 

So as you see, several layer of things going on. If this needs improvement, we need to think about it carefully

**Reply by andrewpbray** [reply]
Thanks @cderv this is helpful!

For now I'm thinking the easiest route will be to write a Lua filter to move `#| id:` from the output to the Div of class cell. Since all of the docs around crossrefs use `#| label:`, hopefully we'll be able to avoid situations where we're breaking or changing that functionality. And if it is too messy, we can switch to using a custom attribute name.

**Reply by andrewpbray** [reply]
Looks like that approach is more brittle than I expected :/

Here is @jimjam-slam 's code block using `id: fig-histplot`:

````
```{r}
#| id: fig-histplot
#| classes: my-histplot

hist(rnorm(5))
```
````
```
      Div
          ( "" , [ "cell" , "my-histplot" ] , [] )
          [ CodeBlock
              ( "" , [ "r" , "cell-code" ] , [] ) "hist(rnorm(5))"
          , Div
              ( "" , [ "cell-output-display" ] , [] )
              [ Plain
                  [ Image
                      ( "fig-histplot" , [] , [] )
                      []
                      ( "demo_files/figure-native/fig-histplot-1.png" , "")
```
Here it is if you don't use a protected prefix:
````
```{r}
#| id: cr-histplot
#| classes: my-histplot

hist(rnorm(5))
```
````
```
      Div
          ( "" , [ "cell" , "my-histplot" ] , [] )
          [ CodeBlock
              ( "" , [ "r" , "cell-code" ] , [] ) "hist(rnorm(5))"
          , Div
              ( "" , [ "cell-output-display" ] , [] )
              [ Para                    <<<<<< Changes from Plain to Para >>>>>>>
                  [ Image
                      ( "" , [] , [] ). <<<<<< id disappears >>>>>>>
                      []
                      ( "demo_files/figure-native/fig-histplot-1.png" , "")
```
It seems like `id` is aliased with `label` and also checks for the prefix before it propagates through to the AST as an identifier. Bummer!

So it seems like, barring any changes to how `id` is handled, our best bet is to use a `custom-id`?

**Reply by cderv** [reply]
That is an interesting difference ! 

`id` and `label` are alias for sure. And we trigger different processing if we find a know prefix in the id. Though I don't know if the `Para` vs `Plain` is on purpose. 

However, we do keep the `label` in our cell processing when it is a known one. otherwise, it is thrown away. This happens in our engine processing - `label` has special meaning  and so `id`

This could be in part related to 
- https://github.com/quarto-dev/quarto-cli/issues/7665

Though I don't think it will help here as id as really a meaning like label. Unless we revert that, or we consider that any label processing should happen in lua and no more at engine level 🤔

**Reply by andrewpbray** [reply]
I had a chat with @cscheid about this and we tinkered with a couple potential fixes - none of them straightforward! It brought up this thread: https://github.com/quarto-dev/quarto-cli/issues/7062 . Even if we solved it for knitr-run cells, Jupyter handles the cell  ids in a different manner (in this case, one that'd work better for our use-case).

Standardizing the way the AST looks across engines sounds very tricky, so for now I think we'll move ahead with wrapping the code cells in a div and labeling it using pandoc.

````markdown

:::{#cr-histplot}
```{r}
hist(rnorm(5))
```
:::
````
